src/include/random.h

```diff
@@ -0,0 +1,389 @@
+{
+  return detail::generate_random_number<RealT,
+                                        std::uniform_real_distribution<RealT>, 
+                                        std::default_random_engine>
+                                       (min, max);
+}
+
+namespace detail {
+
+template <typename RealT, typename MutexT>
+RealT generate_random_number(const RealT max, MutexT& m,
+                             ceph::enable_if_t<std::is_floating_point<RealT>::value>* = nullptr)
+{
+  constexpr RealT zero = 0.0;
+  return generate_random_number(zero, max, m);
+}
+
+template <typename RealT, typename MutexT, typename EngineT>
+RealT generate_random_number(const RealT min, const RealT max, MutexT& m, EngineT& e,
+                             ceph::enable_if_t<std::is_floating_point<RealT>::value>* = nullptr)
+{
+  return detail::generate_random_number<RealT, MutexT, 
+                                        std::uniform_real_distribution<RealT>,
+                                        EngineT>
+                                       (min, max, m, e);
+}
+
+
+template <typename RealT, typename MutexT, typename EngineT>
+RealT generate_random_number(const RealT max, MutexT& m, EngineT& e,
+                             ceph::enable_if_t<std::is_floating_point<RealT>::value>* = nullptr)
+{
+  constexpr RealT zero = 0.0;
+  return generate_random_number(zero, max, m, e);
+}
+
+} // namespace detail
+
+template <typename RealT>
+RealT generate_random_number(const RealT max,
+                             ceph::enable_if_t<std::is_floating_point<RealT>::value>* = nullptr)
+{
+  constexpr RealT zero = 0.0;
+  return generate_random_number(zero, max);
+}
+
+// Convenience functions:
+namespace detail {
+
+// Note that C++11 restricts us here because we can't get at the type of
+// lambda and there's limited return type deduction:
+template <typename IntT,
+          IntT Min, IntT Max, 
+          typename EngineT = std::default_random_engine>
+auto make_random_number_function() -> IntT(*)()
+{
+  return []() {
+             static std::mutex m;
+             static EngineT e;
+
+             return detail::generate_random_number(Min, Max, m, e);
+         };
+}
+
+template <typename IntT,
+          IntT Min, IntT Max, IntT Seed, 
+          typename EngineT = std::default_random_engine>
+auto make_random_number_function() -> IntT(*)()
+{
+  return []() {
+            static std::mutex m;
+            static EngineT e(Seed);
+
+            return detail::generate_random_number(Min, Max, m, e);
+         };
+}
+
+} // namespace detail
+
+template <int Min, int Max, 
+          typename EngineT = std::default_random_engine>
+auto make_random_number_function() -> int(*)()
+{
+  return detail::make_random_number_function<int, Min, Max, EngineT>();
+}
+
+template <int Min, int Max, int Seed,
+          typename EngineT = std::default_random_engine>
+auto make_random_number_function() -> int(*)()
```

Comment:
these function interfaces seem very limited in utility because they only support `int`, and all of the min/max/seed values have to be known at compile time

in what cases would a user want to use `make_random_number_function()` over `random_number_generator`?